model/jpa/src/main/java/org/keycloak/storage/jpa/JpaHashUtils.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2016 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.storage.jpa;
+
+import org.keycloak.crypto.HashException;
+import org.keycloak.crypto.JavaAlgorithm;
+
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.util.Locale;
+import java.util.Objects;
+
+/**
+ * Create hashes for long values stored in the database. Offers different variants for exact and lowercase search.
+ * Keycloak uses lowercase search to approximate a case-insensitive search.
+ * <p>
+ * The lowercase function always uses the English locale to avoid changing hashes due to changing locales which can be surprising
+ * and would be expensive to fix as all hashes would need to be re-calculated.
+ *
+ * @author Alexander Schwartz
+ */
+public class JpaHashUtils {
+
+    private static byte[] hash(byte[] inputBytes) {
+        try {
+            MessageDigest md = MessageDigest.getInstance(JavaAlgorithm.SHA512);
+            md.update(inputBytes);
+            return md.digest();
+        } catch (Exception e) {
+            throw new HashException("Error when creating token hash", e);
+        }
+    }
+
+    public static byte[] hashForAttributeValue(String value) {
+        return JpaHashUtils.hash(value.getBytes(StandardCharsets.UTF_8));
+    }
+
+    public static byte[] hashForAttributeValueLowerCase(String value) {
+        return JpaHashUtils.hash(value.toLowerCase(Locale.ENGLISH).getBytes(StandardCharsets.UTF_8));
+    }
+
+    public static boolean compareSourceValueLowerCase(String value1, String value2) {
+        return Objects.equals(value1.toLowerCase(Locale.ENGLISH), value2.toLowerCase(Locale.ENGLISH));
```

Comment:
A default locale can change during a migration from one system to another, or an upgrade in an unexpected way. With the current setup we can't detect when a locale has changed, and we don't offer a migration path from one locale to another. 

With the setup above, there is at least the guarantee that if you enter the exact same input as before in the same upper/lower case, it will always return that as a match. It might miss out on some variants due to the locale. 

A future version of Keycloak could make this configurable, and could include a re-calculation of the hashes. It would need to iterate over all entries in the table, re-calculate the hashes, and then updating those which differ (which would hopefully be only a subset).

Reply:
I don't think it is feasible to recalculate the hashes in these tables but I don't see any other option. I like the idea of dropping the lowercase search in the future which would solve this.

Thanks for the explanation!